draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -548,6 +592,92 @@ Elligator2 can be simplified with projective coordinates.
 
 ((TODO: write this variant))
 
+## FT-hash
+
+{{FT12}} outline a method for hashing onto a set of curves which includes BN curves, which are suitable
+for bilinear pairings. A number of cryptographic schemes rely on pairings, including some identity-based
+encryption (IBE) and attribute-based encryption (ABE) schemes.  BN curves are not one specific curve but a
+class of curves which can be constructed using Barreto and Naehrig's algorithm {{BN05}}. BN curves are a
+preferred option for asymmetric pairings as they provide relatively efficient computation for pairings, which are
+usually the slowest computation in any cryptosystem using them. There is also work giving parameter selection
+for 256-bit BN-curves, resulting in 128 bits of security for hashing onto BN curves. Suggestions for suitable BN
+curves can be found in {{draft-BN}} (work in progress).
+
+This method is suitable for BN-curves, which have the form E: f(x) = y^2 = x^3 + b, where operations
+are over a field Fp,  where p = 7 (mod 12) and 1 + b is a nonzero square in Fp. Fouque and Tibouchi give an 
+encoding F which makes the hash function Hash(m) = F(H1(m)) + F(H2(m)) indifferentiable from a random oracle,
+where  H1 and H2 are independent hash functions into Fp. We give this encoding and hash function construction
+in the steps below.  In the following, Xp() is the Legendre symbol (Xp(0) = 0, for a != 0 Xp(a) = 1 if a is a square in
+Fp and -1 otherwise), which can be computed by Xp(a) = a ^ ((p -1) / 2).
+
+~~~
+FT-encoding(t)
+
+Input:
+
+  t - the input, an element in Fp
+
+  f() - Curve function
+
+Output:
+
+  (x, y) - a point in E
+
+Steps:
+
+1.   s = (-3)^(1/2) (mod p)
+2.   x1 = (-1 + s)/2  -  (s t^2) / (1 + b + t^2) (mod p)
+3.   x2 = (-1 - s)/2  +  (s t^2) / (1 + b + t^2) (mod p)
+4.   x3 = 1 - ((1 + b + t^2)^2)  /  (3 t^2) (mod p)
+5.   s1 = Xp(g(x1))  // calculate quadratic residue of g(x1)
+6.   s2 = Xp(g(x2))  // calculate quadratic residue of g(x2)
+7.   s3 = Xp(g(x3))  // calculate quadratic residue of g(x3)
+8.   if s1 = 1: x = x1  // choose the smallest index i such that g(xi) is a square in Fp
```

Comment:
We should re-write this conditional selection to make it constant time, too, right?

Reply:
This should ideally be addressed, something like:
```
8.  x = x3
9.  x = CMOV(x2 ,x, 2)
10. x = CMOV(x1, x, s1)
```